reporter.go

```diff
@@ -0,0 +1,302 @@
+	Time     float64  `xml:"time,attr"`
+	Failures []jUnitTestReportFailure
+}
+
+type jUnitReportSchema struct {
+	XMLName   xml.Name `xml:"testsuite"`
+	Name      string   `xml:"name,attr"`
+	Tests     int      `xml:"tests,attr"`
+	Errors    int      `xml:"errors,attr"`
+	Failures  int      `xml:"failures,attr"`
+	Time      float64  `xml:"time,attr"`
+	TestCases []jUnitTestReportSchema
+}
+
+func createJunitReportSchema(runDetails *rainforest.RunDetails, client *rainforest.Client) (*jUnitReportSchema, error) {
+	finalStateName := runDetails.StateDetails.Name
+	duration, err := timeStringDifferenceSecs(runDetails.Timestamps["created_at"], runDetails.Timestamps[finalStateName])
+	if err != nil {
+		return &jUnitReportSchema{}, err
+	}
+
+	var testCases *[]jUnitTestReportSchema
+	testCases, err = createJunitTestReportSchema(runDetails.ID, &runDetails.Tests, client)
+	if err != nil {
+		return &jUnitReportSchema{}, err
+	}
+
+	report := &jUnitReportSchema{
+		Name:      runDetails.Description,
+		Errors:    runDetails.TotalNoResultTests,
+		Failures:  runDetails.TotalFailedTests,
+		Tests:     runDetails.TotalTests,
+		TestCases: *testCases,
+		Time:      duration,
+	}
+
+	return report, nil
+}
+
+func createJunitTestReportSchema(runID int, tests *[]rainforest.RunTestDetails, api reporterAPI) (*[]jUnitTestReportSchema, error) {
+	type processedTestCase struct {
+		TestCase jUnitTestReportSchema
+		Error    error
+	}
+
+	testChan := make(chan processedTestCase, len(*tests))
+
+	// Limit the maximum concurrent requests
+	httpThreads := make(chan struct{}, reporterConcurrency)
+
+	processTest := func(test rainforest.RunTestDetails) {
+		testCase := jUnitTestReportSchema{}
+
+		duration, err := timeStringDifferenceSecs(test.CreatedAt, test.UpdatedAt)
+
+		if err != nil {
+			testChan <- processedTestCase{TestCase: testCase, Error: err}
+			return
+		}
+
+		testCase.Name = test.Title
+		testCase.Time = duration
+
+		if test.Result == "failed" {
+			var testDetails *rainforest.RunTestDetails
+
+			// reserve a thread
+			httpThreads <- struct{}{}
+			log.Printf("Fetching information for failed test #" + strconv.Itoa(test.ID))
+			testDetails, err = api.GetRunTestDetails(runID, test.ID)
+			// release the thread
+			<-httpThreads
+
+			if err != nil {
+				testChan <- processedTestCase{TestCase: jUnitTestReportSchema{}, Error: err}
+				return
+			}
+
+			for _, step := range testDetails.Steps {
+				for _, browser := range step.Browsers {
+					browserName := browser.Name
+
+					for _, feedback := range browser.Feedback {
+						if feedback.AnswerGiven == "no" && feedback.JobState == "approved" && feedback.Note != "" {
+							reportFailure := jUnitTestReportFailure{Type: browserName, Message: feedback.Note}
+							testCase.Failures = append(testCase.Failures, reportFailure)
+						}
+					}
+				}
+			}
+		}
+
+		testChan <- processedTestCase{TestCase: testCase}
+	}
+
+	for _, test := range *tests {
+		go processTest(test)
+	}
```

Comment:
One thing to think about - I'm a fan of worker pool pattern for things like this. So instead of spawning sub routines for each process and then halting them if the rate limit is reached, spawn limited number of workers that will go through the data - as in tabular vars concurrency.  Although it's also fine way it is, so that's just something to think about. ;)